File: src/core/castImage.ts
import Image from './Image.js'
import CastImageOptions from './CastImageOptions.js'
import PixelTypes from './PixelTypes.js'
import IntTypes from './IntTypes.js'
import FloatTypes from './FloatTypes.js'

/**
 * Cast an image to another PixelType or ComponentType
 *
 * @param {Image} image - The input image
 * @param {CastImageOptions} options - specify the componentType and/or pixelType of the output
 */
function castImage (inputImage: Image, options?: CastImageOptions): Image {
  const outputImageType = { ...inputImage.imageType }

  if (typeof options !== 'undefined' && typeof options.pixelType !== 'undefined') {
    outputImageType.pixelType = options.pixelType
    if (options.pixelType === PixelTypes.Scalar && outputImageType.components !== 1) {
      throw new Error('Cannot cast multi-component image to a scalar image')
    }
  }
  if (typeof options !== 'undefined' && typeof options.componentType !== 'undefined' && options.componentType !== inputImage.imageType.componentType) {
    outputImageType.componentType = options.componentType
  }

  const outputImage = new Image(outputImageType)

  outputImage.name = inputImage.name
  outputImage.origin = Array.from(inputImage.origin)
  outputImage.spacing = Array.from(inputImage.spacing)
  outputImage.direction = inputImage.direction.slice()
  outputImage.size = Array.from(inputImage.size)
  // Deep copy the map
  outputImage.metadata = new Map(JSON.parse(JSON.stringify(Array.from(inputImage.metadata))))

  if (inputImage.data !== null) {
    if (typeof options !== 'undefined' && typeof options.componentType !== 'undefined' && options.componentType !== inputImage.imageType.componentType) {
      switch (inputImage.imageType.componentType) {
        case IntTypes.UInt8:
        case IntTypes.Int8:
        case IntTypes.UInt16:
        case IntTypes.Int16:
        case IntTypes.UInt32:
        case IntTypes.Int32:
        case FloatTypes.Float32:
        case FloatTypes.Float64:
          switch (outputImage.imageType.componentType) {
            case IntTypes.UInt8:
              outputImage.data = new Uint8Array(inputImage.data)
              break
            case IntTypes.Int8:
              outputImage.data = new Int8Array(inputImage.data)
              break
            case IntTypes.UInt16:
              outputImage.data = new Uint16Array(inputImage.data)
              break
            case IntTypes.Int16:
              outputImage.data = new Int16Array(inputImage.data)
              break
            case IntTypes.UInt32:
              outputImage.data = new Uint32Array(inputImage.data)
              break
            case IntTypes.Int32:
              outputImage.data = new Int32Array(inputImage.data)
              break
            case FloatTypes.Float32:
              outputImage.data = new Float32Array(inputImage.data)
              break
            case FloatTypes.Float64:
              outputImage.data = new Float64Array(inputImage.data)
              break
            case IntTypes.UInt64:
              outputImage.data = new BigUint64Array(inputImage.data.length)
              for (let idx = 0; idx < outputImage.data.length; idx++) {
                outputImage.data[idx] = BigInt.asIntN(64, BigInt(inputImage.data[idx]))
              }
              break
            case IntTypes.Int64:
              outputImage.data = new BigInt64Array(inputImage.data.length)
              for (let idx = 0; idx < outputImage.data.length; idx++) {
                outputImage.data[idx] = BigInt.asUintN(64, BigInt(inputImage.data[idx]))
              }
              break
          }
          break
        case IntTypes.UInt64:
        case IntTypes.Int64:
          switch (outputImage.imageType.componentType) {
            case IntTypes.UInt8:
              outputImage.data = new Uint8Array(inputImage.data.length)
              for (let idx = 0; idx < outputImage.data.length; idx++) {
                outputImage.data[idx] = Number(inputImage.data[idx])
              }
              break
            case IntTypes.Int8:
              outputImage.data = new Int8Array(inputImage.data.length)
              for (let idx = 0; idx < outputImage.data.length; idx++) {
                outputImage.data[idx] = Number(inputImage.data[idx])
              }
              break
            case IntTypes.UInt16:
              outputImage.data = new Uint16Array(inputImage.data.length)
              for (let idx = 0; idx < outputImage.data.length; idx++) {
                outputImage.data[idx] = Number(inputImage.data[idx])
              }
              break
            case IntTypes.Int16:
              outputImage.data = new Int16Array(inputImage.data.length)
              for (let idx = 0; idx < outputImage.data.length; idx++) {
                outputImage.data[idx] = Number(inputImage.data[idx])
              }
              break
            case IntTypes.UInt32:
              outputImage.data = new Uint32Array(inputImage.data.length)
              for (let idx = 0; idx < outputImage.data.length; idx++) {
                outputImage.data[idx] = Number(inputImage.data[idx])
              }
              break
            case IntTypes.Int32:
              outputImage.data = new Int32Array(inputImage.data.length)
              for (let idx = 0; idx < outputImage.data.length; idx++) {
                outputImage.data[idx] = Number(inputImage.data[idx])
              }
              break
            case FloatTypes.Float32:
              outputImage.data = new Float32Array(inputImage.data.length)
              for (let idx = 0; idx < outputImage.data.length; idx++) {
                outputImage.data[idx] = Number(inputImage.data[idx])
              }
              break
            case FloatTypes.Float64:
              outputImage.data = new Float64Array(inputImage.data.length)
              for (let idx = 0; idx < outputImage.data.length; idx++) {
                outputImage.data[idx] = Number(inputImage.data[idx])
              }
              break
            case IntTypes.UInt64:
              outputImage.data = new BigUint64Array(inputImage.data)
              break
            case IntTypes.Int64:
              outputImage.data = new BigInt64Array(inputImage.data)
              break
          }
          break
      }
    } else {
      // copy
      const CTor = inputImage.data.constructor as new(length: number) => typeof inputImage.data
      outputImage.data = new CTor(inputImage.data.length)
      if (outputImage.data != null) {
        // @ts-expect-error: error TS2345: Argument of type 'TypedArray' is not assignable to parameter of type 'ArrayLike<number> & ArrayLike<bigint>'
        outputImage.data.set(inputImage.data, 0)
      }
    }
  }

  return outputImage
}

export default castImage

File: packages/core/python/itkwasm/itkwasm/cast_image.py
from typing import Optional, Union
from dataclasses import asdict
import copy

import numpy as np

from .image import Image, ImageType
from .pixel_types import PixelTypes
from .int_types import IntTypes
from .float_types import FloatTypes


def cast_image(
    input_image: Image,
    pixel_type: Optional[PixelTypes] = None,
    component_type: Optional[Union[IntTypes, FloatTypes]] = None,
) -> Image:
    """Cast an image to another pixel type and / or component type.

    :param input_image: Image to be cast.
    :type  input_image: Image

    :param pixel_type: Pixel type to cast to.
    :type  pixel_Type: Optional[PixelTypes]

    :param component_type: Component type to cast to.
    :type  component_type: Optional[Union[IntTypes, FloatTypes]]

    :return: Cast image. Event if pixel_type or component_type are not specified, a copy is made.
    :rtype:  Image

    """
    output_image_type = ImageType(**asdict(input_image.imageType))

    if pixel_type is not None:
        output_image_type.pixelType = pixel_type
        if pixel_type == PixelTypes.Scalar and output_image_type.components != 1:
            raise ValueError("PixelType Scalar requires components == 1")

    if component_type is not None and component_type != input_image.imageType.componentType:
        output_image_type.componentType = component_type

    output_image = Image(output_image_type)

    output_image.name = input_image.name
    output_image.origin = list(input_image.origin)
    output_image.spacing = list(input_image.spacing)
    output_image.direction = input_image.direction.copy()
    output_image.size = list(input_image.size)
    output_image.metadata = copy.deepcopy(input_image.metadata)

    if input_image.data is not None:
        if output_image_type.componentType == input_image.imageType.componentType:
            output_image.data = input_image.data.copy()
        else:
            component_type = output_image_type.componentType
            if component_type == IntTypes.UInt8:
                output_image.data = input_image.data.astype(np.uint8)
            elif component_type == IntTypes.Int8:
                output_image.data = input_image.data.astype(np.int8)
            elif component_type == IntTypes.UInt16:
                output_image.data = input_image.data.astype(np.uint16)
            elif component_type == IntTypes.Int16:
                output_image.data = input_image.data.astype(np.int16)
            elif component_type == IntTypes.UInt32:
                output_image.data = input_image.data.astype(np.uint32)
            elif component_type == IntTypes.Int32:
                output_image.data = input_image.data.astype(np.int32)
            elif component_type == IntTypes.UInt64:
                output_image.data = input_image.data.astype(np.uint64)
            elif component_type == IntTypes.Int64:
                output_image.data = input_image.data.astype(np.int64)
            elif component_type == FloatTypes.Float32:
                output_image.data = input_image.data.astype(np.float32)
            elif component_type == FloatTypes.Float64:
                output_image.data = input_image.data.astype(np.float64)
            else:
                raise ValueError("Unsupported component type")

    return output_image
